text/0000-optimize-by-default.md

```diff
@@ -0,0 +1,166 @@
+and to include debug info.
+
+Debate is welcome on the specifics as long as "unadorned" `rustc` and `cargo
+build` calls produce optimized binaries.
+
+# Drawbacks
+
+## Slower builds by default
+
+Builds will only be slower _by default_; in other words, anyone calling `rustc
+foo.rs` today can trivially call `rustc -C optlevel=0 foo.rs` or `rustc --debug
+foo.rs` if the extra `rustc` flags are implemented.
+
+Note the failure state of debug-by-default: if you wanted an optimized build
+instead, your production binary is now incredibly slow. As someone who has
+actually pushed unoptimized binaries to production by accident, I can
+confidently say _this costs a lot of money._
+
+For opt-by-default, if you actually wanted a debug build the failure state is
+"you waited longer for your build to finish than you should have," which while
+annoying, has _far_ less potential for damage.
+
+## Breaks convention with compilers for other languages
+
+While this is a real concern, it's the result of those compilers continuing to
+carry a legacy burden. C compilers need to be backwards compatible with the
+state-of-the-art from many decades ago, and C++ compilers need to be compatible
+with C compilers. In other words, they don't have a choice.
+
+Rust has been breaking plenty of language conventions for the sake of safety; it
+should continue on this commendable path without being hamstrung by decisions
+made in 1972.
+
+# Alternatives
+
+## Only have `cargo` echo a message stating `debug` or `release` build mode
+
+This has been recently talked about, and while it's a welcome and useful
+feature, it sadly isn't enough. A Rust newcomer who's only used Python, Ruby,
+JavaScript, Java, C# or a similar language is not going to understand the
+consequences of a "building in debug mode" message.
+
+Not to mention that it's easy to miss such messages when `cargo build` is run by
+a script that will also call plenty of other commands and will thus have lots of
+output.
+
+Not to mention it does nothing for `rustc` calls by newbies. See the next
+section for details.
+
+## Only make `cargo` opt-by-default, but leave `rustc` as is
+
+While this alone would be a large safety & usability boost over the current
+state, it would not help Rust newcomers or those playing around with/evaluating
+the language. Such users are less likely to know about `cargo`, and even if they
+do, are likely to shun it while experimenting with simple programs. In such
+cases, they could easily conclude that "Rust is too slow" based on their initial
+experiments and never reach `cargo build`.
+
+`rustc foo.rs` is understandably very tempting when trying to wrap one's head around the
+language and it would (like today) speak badly about Rust's performance.
+
+## Make `rustc` include debug info along with optimized code
+
+This is better than the previous alternative, but would still lead to questions
+about small Rust programs leading to huge binaries.
```

Comment:
> debug info in optimised code is often far less useful than that in unoptimised (or "lightly" optimised) code.

I'm aware of this, but some people do prefer debug info in opt binaries (at least in the C++ world) so I wanted to mention this alternative.